File: 81. LowestCommonAncestorBinaryTree.c
#include <stdio.h>
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct TreeNode *newNode(int val) {
    struct TreeNode *node = malloc(sizeof(*node));
    node->val = val;
    node->left = node->right = NULL;
    return node;
}

struct TreeNode *lowestCommonAncestor(struct TreeNode *root, struct TreeNode *p, struct TreeNode *q) {
    if (!root || root == p || root == q) return root;
    struct TreeNode *left  = lowestCommonAncestor(root->left,  p, q);
    struct TreeNode *right = lowestCommonAncestor(root->right, p, q);
    if (left && right) return root;
    return left ? left : right;
}

int main(void) {
    struct TreeNode *root = newNode(3);
    root->left = newNode(5);
    root->right = newNode(1);
    root->left->left = newNode(6);
    root->left->right = newNode(2);
    root->right->left = newNode(0);
    root->right->right = newNode(8);
    root->left->right->left  = newNode(7);
    root->left->right->right = newNode(4);

    // lca1: LCA of node 5 and node 1 => 3
    struct TreeNode *lca1 = lowestCommonAncestor(root, root->left, root->right);
    // lca2: LCA of node 5 and node 4 => 5
    struct TreeNode *lca2 = lowestCommonAncestor(root, root->left, root->left->right->right);

    printf("%d\n", lca1->val);
    printf("%d\n", lca2->val);
    return 0;
}

File: 81. LowestCommonAncestorBinaryTree.java
class TreeNodeLCA {
    int val;
    TreeNodeLCA left;
    TreeNodeLCA right;

    TreeNodeLCA(int val) {
        this.val = val;
    }
}

class Solution {
    public TreeNodeLCA lowestCommonAncestor(TreeNodeLCA root, TreeNodeLCA p, TreeNodeLCA q) {
        if (root == null || root == p || root == q) {
            return root;
        }

        TreeNodeLCA left = lowestCommonAncestor(root.left, p, q);
        TreeNodeLCA right = lowestCommonAncestor(root.right, p, q);

        if (left != null && right != null) {
            return root;
        }

        return left != null ? left : right;
    }
}

public class LowestCommonAncestorBinaryTree {
    public static void main(String[] args) {
        TreeNodeLCA root = new TreeNodeLCA(3);
        root.left = new TreeNodeLCA(5);
        root.right = new TreeNodeLCA(1);
        root.left.left = new TreeNodeLCA(6);
        root.left.right = new TreeNodeLCA(2);
        root.right.left = new TreeNodeLCA(0);
        root.right.right = new TreeNodeLCA(8);
        root.left.right.left = new TreeNodeLCA(7);
        root.left.right.right = new TreeNodeLCA(4);

        Solution solution = new Solution();

        TreeNodeLCA lca1 = solution.lowestCommonAncestor(root, root.left, root.right);
        TreeNodeLCA lca2 = solution.lowestCommonAncestor(root, root.left, root.left.right.right);

        System.out.println(lca1.val);
        System.out.println(lca2.val);
    }
}
